esmerald/encoders.py

```diff
@@ -1,102 +1,70 @@
-
-
-class MsgSpecEncoder(Encoder):
-    def is_type(self, value: Any) -> bool:
-        return isinstance(value, Struct) or is_class_and_subclass(value, Struct)
-
-    def serialize(self, obj: Any) -> Any:
-        """
-        When a `msgspec.Struct` is serialised,
-        it will call this function.
-        """
-        return msgspec.json.decode(msgspec.json.encode(obj))
-
-    def encode(self, annotation: Any, value: Any) -> Any:
-        return msgspec.json.decode(msgspec.json.encode(value), type=annotation)
+def register_esmerald_encoder(
+    encoder: EncoderProtocol | MoldingProtocol | type[EncoderProtocol] | type[MoldingProtocol],
+) -> None:
+    """
+    Registers an esmerald encoder into available Lilya encoders
+    """
+    try:
+        register_encoder(encoder)
+    except RuntimeError:
+        raise ImproperlyConfigured(f"{type(encoder)} must be a subclass of Encoder") from None
 
 
-class PydanticEncoder(Encoder):
-    def is_type(self, value: Any) -> bool:
-        return isinstance(value, BaseModel) or is_class_and_subclass(value, BaseModel)
+PydanticEncoder = ModelDumpEncoder
+try:
+    import msgspec
 
-    def serialize(self, obj: BaseModel) -> dict[str, Any]:
-        try:
-            return obj.model_dump(mode="json")
-        except PydanticSerializationError:
-            return obj.model_dump()
+    class MsgSpecEncoder(Encoder):
+        __type__ = msgspec.Struct
 
-    def encode(self, annotation: Any, value: Any) -> Any:
-        if isinstance(value, BaseModel) or is_class_and_subclass(value, BaseModel):
-            return value
-        return annotation(**value)
+        def serialize(self, obj: Any) -> Any:
+            """
+            When a `msgspec.Struct` is serialised,
+            it will call this function.
+            """
+            return msgspec.json.decode(msgspec.json.encode(obj))
 
+        def encode(self, annotation: Any, value: Any) -> Any:
+            return msgspec.json.decode(msgspec.json.encode(value), type=annotation)
 
-def register_esmerald_encoder(encoder: Encoder[Any]) -> None:
-    """
-    Registers an esmerald encoder into available Lilya encoders
-    """
-    if not isinstance(encoder, Encoder) and not is_class_and_subclass(encoder, Encoder):  # type: ignore
-        raise ImproperlyConfigured(f"{type(encoder)} must be a subclass of Encoder")
+    register_esmerald_encoder(MsgSpecEncoder)
+except ImportError:
+    pass
 
-    encoder_types = {encoder.__class__.__name__ for encoder in ENCODER_TYPES}
-    if encoder.__name__ not in encoder_types:
-        register_encoder(encoder)
+json_encoder = json_encode
 
 
 def is_body_encoder(value: Any) -> bool:
     """
     Function that checks if the value is a body encoder.
     """
+    encoder_types = ENCODER_TYPES_CTX.get()
     if not is_union(value):
-        return any(encoder.is_type(value) for encoder in ENCODER_TYPES)
+        return any(encoder.is_type(value) for encoder in encoder_types)
```

Comment:
I want this like it was before. ENCODER_TYPES = LILYA_ENCODER_TYPES.get() at the top and its used here directly. No need for reassignments, at all.